app/model/blockchain/token.py

```diff
@@ -16,21 +16,66 @@
 
 SPDX-License-Identifier: Apache-2.0
 """
+from __future__ import annotations
 
 import json
-from decimal import Decimal
 from datetime import datetime, timedelta
-
+from decimal import Decimal
 from eth_utils import to_checksum_address
+import functools
+from typing import Union, Type, Optional, Callable
+from sqlalchemy.orm import Session
 
 from app import config
 from app import log
 from app.contracts import Contract
-from app.model.db import Listing
+from app.model.db import (
+    Listing,
+    IDXBondToken as BondTokenModel,
+    IDXShareToken as ShareTokenModel,
+    IDXMembershipToken as MembershipTokenModel,
+    IDXCouponToken as CouponTokenModel,
+)
+from app.model.db.idx_token import TokenModelClassTypes, TokenModelInstanceTypes
 from app.utils.company_list import CompanyList
 
 LOG = log.get_logger()
 
+TokenClassTypes = Union[Type['BondToken'], Type['ShareToken'], Type['MembershipToken'], Type['CouponToken']]
+TokenInstanceTypes = Union['BondToken', 'ShareToken', 'MembershipToken', 'CouponToken']
+
+
+def token_db_cache(TargetModel: TokenModelClassTypes):
+    """Cache decorator for Token Details"""
+
+    def decorator(func: Callable[[TokenClassTypes, Session, str], Optional[TokenInstanceTypes]]):
+        @functools.wraps(func)
+        def wrapper(cls: TokenClassTypes, session: Session, token_address: str) -> Optional[TokenInstanceTypes]:
+            if not config.TOKEN_CACHE:
+                return func(cls, session, token_address)
+
+            # Get data from cache
+            cached_token: Optional[TokenModelInstanceTypes] = \
+                session.query(TargetModel).filter(TargetModel.token_address == token_address).first()
+
+            if cached_token and cached_token.modified < datetime.utcnow() + timedelta(seconds=config.TOKEN_CACHE_TTL):
+                return cls.from_model(cached_token)
+
+            # Get data from blockchain node
+            result = func(cls, session, token_address)
+
+            if result:
+                # Save data to cache
+                new_cache = result.to_model()
+                if cached_token:
+                    session.merge(new_cache)
+                else:
+                    session.add(new_cache)
+            return result
+
+        return wrapper
+    return decorator
+
```

Comment:
Added db cache decorator for token details so that another cache decorator(e.g. process memory) also can be used in same time.